File: ocaml/functional/sorting.ml
type sort_data_type = StringType | DoubleType


let read_file_lines filename =
  let ic = open_in filename in
  let rec read_lines acc =
    try read_lines (input_line ic :: acc) with End_of_file -> close_in ic; List.rev acc
  in
  read_lines []

let write_file_lines filename lines =
  let oc = open_out filename in
  List.iter (fun line -> Printf.fprintf oc "%s\n" line) lines;
  close_out oc

  let is_double line =
    try
      let _ = float_of_string line in true
    with Failure _ -> false
  
  let determine_data_type lines =
    List.fold_left (fun acc line ->
      if acc = StringType then StringType
      else if is_double line then DoubleType
      else StringType
    ) DoubleType lines
  

let rec quicksort = function
  | [] -> []
  | pivot :: rest ->
      let less, greater = List.partition (fun x -> x <= pivot) rest in
      quicksort less @ [pivot] @ quicksort greater

let process_dataset input output =
  let lines = read_file_lines input in
  let _ = determine_data_type lines in
  let sorted = quicksort lines in
  write_file_lines output sorted

let () =
  process_dataset "datasets/random_char_strings.csv" "./sorted_lines_string.txt";
  process_dataset "datasets/random_double_floats.csv" "./sorted_lines_double.txt"

File: rust/functional/sorting.rs
use std::fs::File;
use std::io::{BufRead, BufReader, BufWriter, Write};

#[derive(Debug, PartialEq)]
enum SortDataType { StringType, DoubleType }

// Read lines from a file (functional style)
fn read_file_lines(filename: &str) -> Vec<String> {
    BufReader::new(File::open(filename).unwrap())
        .lines()
        .map(|l| l.unwrap())
        .collect()
}

// Write lines to a file (functional style)
fn write_file_lines(filename: &str, lines: &[String]) {
    BufWriter::new(File::create(filename).unwrap())
        .write_all(lines.join("\n").as_bytes())
        .unwrap();
}

// Check if all lines are doubles using parse::<f64>()
fn determine_data_type(lines: &[String]) -> SortDataType {
    if lines.iter().all(|line| line.parse::<f64>().is_ok()) {
        SortDataType::DoubleType
    } else {
        SortDataType::StringType
    }
}

// Quicksort with iterators and recursion
fn quicksort(lines: &[String]) -> Vec<String> {
    match lines {
        [] | [_] => lines.to_vec(),
        [pivot, rest @ ..] => {
            let (less, greater): (Vec<String>, Vec<String>) = rest.iter()
                .cloned()
                .partition(|line| line <= pivot);
            [quicksort(&less), vec![pivot.clone()], quicksort(&greater)].concat()
        }
    }
}

fn main() {
    // Process datasets
    let process_dataset = |input: &str, output: &str| {
        let lines = read_file_lines(input);
        let sorted = quicksort(&lines);
        write_file_lines(output, &sorted);
    };

    process_dataset("datasets/random_char_strings.csv", "./sorted_lines_string.txt");
    process_dataset("datasets/random_double_floats.csv", "./sorted_lines_double.txt");
}